Add custom type annotations

sakshamphul opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section
This PR adds desired custom type annotation that checks for required columns and maps input columns to output columns using config file

#### What feature is being added or bug is being addressed?

Ability of fusion standardization function to accept custom type input that checks for few required columns. Config file has been made optionally and can be use to map input columns with output columns. Additionally, it can be used to generate output columns with specific header name and 'NA' as an entry (Usage: `"None" :: "Output_name"`)


#### What was your approach?
Developed custom type on top of existing starfusion and arriba. Package will continue to support both of these caller types with same implementation.

#### What GitHub issue does your pull request address?
https://app.zenhub.com/workspaces/d3b-bfx-5cdd720eed0dce4209e0b8a5/issues/d3b-center/bixu-tracker/1533

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.
All the unit test were successful related with fusion standardization function. A new test was also added for custom type

#### Which areas should receive a particularly close look?
fusion standardization function

#### Is there anything that you want to discuss further?
Testing this version of the package will require installing annoFuse locally. Here are the steps to install it locally:
1) Git clone this branch
2) Uninstall previous version of annoFuse `remove.packages("annoFuse")` 
3) Install local version `devtools::install("path/to/package/folder")`
4) Now, you should be able to test fusion_standardization function. 

#### Documentation Checklist

<!-- Please review and specify if it isn't applicable -->

- [X] The function has examples to showcase the usage 
- [ ] The function is fully documented in the package using roxygen 
- [ ] Added a vignette

#### Package Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [X] The dependencies required for the function in this pull request have been added to [DESCRIPTION](https://github.com/d3b-center/annoFuse/blob/master/DESCRIPTION) `Imports` section, and the corresponding `NAMESPACE` directives have been added/updated
- [X] The files needed for the function to run are in `inst/extdata`
- [X] The code was run through `styler`

Commit:
First commit to add custom type annotations

sakshamphul committed:
Update test

sakshamphul committed:
Add config file

sakshamphul committed:
Clean fusion stand file

sickler-alex reviewed (commented):
Overall looks good, I'll test this with what I'm doing in the toolkit to make sure it's still producing the same output on star and arriba files. Can you run Prettier on the scripts you edited to better format the code? If you're using VScode, this is the link to the extension. https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

federicomarini commented:
As an alternative to Prettier, there's the `styler` package which already implements most of the style guidelines I (try to) follow :)
Content-focused review will follow 😉

sickler-alex commented:
I was able to test the updated code with star and arriba files and they produce the same output as previous versions which is expected.

federicomarini committed:
updating RoxygenNote in the DESCRIPTION file

federicomarini committed:
uniforming to the commonly used tags for the manpage documentation

sakshamphul committed:
Run through styler

sakshamphul committed:
Run through styler

federicomarini committed:
Merge branch 'Add-Custom-Type-Annotation' of github.com:d3b-center/annoFuse into Add-Custom-Type-Annotation

federicomarini committed:
cleaning up the R folder from data example debris, now in annoFuseData

federicomarini committed:
cleaning up the R folder from data example debris, now in annoFuseData

federicomarini committed:
rerendering manpage

federicomarini committed:
rerendering manpage

federicomarini committed:
commenting out the portions in the example that are not code to run

federicomarini committed:
commenting out the portions in the example that are not code to run

federicomarini committed:
adjusting indentation

federicomarini committed:
adjusting indentation

federicomarini committed:
fixing indentation

federicomarini committed:
fixing indentation

federicomarini committed:
removing DS_Store file - not required in the repo

federicomarini committed:
removing DS_Store file - not required in the repo

federicomarini commented on R/fusion_standardization.R:
```diff
@@ -29,15 +32,33 @@
 #'   tumorID = "tumorID"
 #' )
+#' # read in CUSTOM type file
+#' formattedCUSTOM <- fusion_standardization(fusionfileStarFusion,
+#'   caller = "CUSTOM",
+#'   tumorID = "All",
+#'   input_json_file = system.file("extdata", "config", package = "annoFuseData")
+#' )
```
this example seems to be not working correctly - can you please have a look at this @sakshamphul ?
Probably, accordingly, we might need to adjust the lines in the unit tests?

sakshamphul replied:
@federicomarini  Can you please describe the issue so that I can look into it? Did you had an error pulling the config file from annoFuseData package?

sakshamphul replied:
@federicomarini Regarding your comment on warnings, I had few print statements in. Can you share the warning  while running the custom type ?

federicomarini replied:
About the example: The error is triggered by simply running the code in the @examples section.
When running R CMD check, this is done automatically, so I get something like this:

<img width="689" alt="image" src="https://user-images.githubusercontent.com/7447478/191264587-f58e97a6-d7b4-4f43-b91c-e8d06efd7d8d.png">

sakshamphul replied:
Aha I see! We have this error because code is unable to find config file in annoFusedata package. We haven't released the latest version of annoFusedata package to CRAN server. I believe once we release annoFusedata package to CRAN, this will be fixed

federicomarini replied:
Hmm, not really I guess. I do have the latest version of annoFuseData installed, and it is correctly picked.
Otherwise I would have had a different error message - the one telling me there is no file with the specs in it.

To me, the culprit seems to be that the fusionfileStarFusion does not have the columns required.
Also, from the live run of the lines:

<img width="664" alt="image" src="https://user-images.githubusercontent.com/7447478/191267206-7eb15c7e-50dd-4acd-90c1-b1ea4d760ca3.png">

I can suggest you use `message` instead of `print` for these statements

```
[1] "Annotating based on config file"
[1] "All required columns exists!"
```

federicomarini replied:
nvm, the print->message has just been done :)

federicomarini replied:
(but the error above stays)

sakshamphul replied:
@federicomarini  Can you also share input data to this function ?

federicomarini replied:
I am running the example from top to bottom, simply that.
Like in "start R, load annoFuse, execute this series of commands line by line"

federicomarini replied:
(which is exactly what R CMD check does to do all the checks and see if this works "just like we said")

sakshamphul replied:
I run it and was unable to reproduce the error 
```
df <- read_tsv("custom_fusions_example.tsv", col_names = TRUE,show_col_types = FALSE)
dataframe_custom <-as.data.frame(df)
output=fusion_standardization(dataframe_custom,"CUSTOM","ALL",input_json_file = system.file("extdata", "config", package = "annoFuseData"))

federicomarini replied:
Well, that is different from what I had to run in the example 😉 
In the current status you would still need to define (and potentially put in the annoFuseData an example for that) the `fusionfileCustom` path. Otherwise, I would (correctly) get an error
<img width="668" alt="image" src="https://user-images.githubusercontent.com/7447478/191294023-977622e9-d979-4b96-add2-a3fd85e05a5a.png">

Actually, you can also create that data.frame on the fly in the example, if it makes sense to do so.
That's pretty much up to you 😃

sakshamphul replied:
Starfusion and arriba annotation requires input data as an data.frame. Just to be consistent with previous methods, I kept it as data.frame for custom type

federicomarini replied:
Sure, I'm not debating that. It is only that you need a "real example" of a mini data frame that gets passed in this step. Otherwise the check of the example is going to return an error.

Just do something like 

```
fusionfileCustom <- data.frame(
  col1_you_need = c(,,,,),
  col2_you_need = c(,,,,),
  col3_you_need = c(,,,,)
)
```
... in the lines before the call to `fusion_standardization`. I am putting a placeholder in the code

federicomarini committed:
introducing some better indentation in the test suite

federicomarini committed:
introducing some better indentation in the test suite

federicomarini commented on tests/testthat/test-fusion_standardization.R:
```diff
@@ -1,6 +1,7 @@
 context("Test fusion standardization")
 
```
I brought in some indentation for easier reading of the code

federicomarini reviewed (changes requested):
The examples seem to have an error in the last portion - for the custom format.

In general: when running the code for the existing functions, some warnings are also triggered - can we try to have them not showing?

Otherwise, the output of R CMD check does not point out to any major changes required.

federicomarini committed:
replacing print with message in info returned while running the function

federicomarini committed:
replacing print with message in info returned while running the function

federicomarini committed:
re-rendering manpage

federicomarini committed:
re-rendering manpage

sakshamphul committed:
Update the example for custom type

sakshamphul committed:
Update the example for custom type

federicomarini committed:
added placeholder

federicomarini committed:
added placeholder

sakshamphul committed:
Fill placeholder

sakshamphul committed:
Fill placeholder

federicomarini committed:
Fixing width of the example line, simplifying readability of the three entries

federicomarini commented:
Well done, the examples run correctly now!

federicomarini commented:
One last thought before the full approval: Please consider if we might need somewhere "outside" the new main function the `shape_output` function. If that is the case, just proceed to document it fully via roxygen and so.
Actually, even if it is not exported, it is good practice (= "would never hurt") to keep it fully documented even if not exported. This makes the whole debugging-upon-changes much easier 😉

sakshamphul committed:
improve the documentation and address the feedback

sakshamphul committed:
improve the documentation and address the feedback

sakshamphul commented:
> 
Ok, I added few comments for shape_output function and updated the documentation via roxygen

federicomarini committed:
Merge branch 'Add-Custom-Type-Annotation' of github.com:d3b-center/annoFuse into Add-Custom-Type-Annotation

federicomarini committed:
rendering manpage for shape_output

federicomarini commented:
Seems all good to me now 😉 
If you need the formal approval of someone else, just ping them. Otherwise, looks like it is ready to merge.

migbro reviewed (dismissed):
I will also approve given the extensive testing this has undergone. I am already familiar with the intent of the update. Many thanks to @federicomarini for your comments and contributions

sakshamphul committed:
Recover the last commit

federicomarini committed:
fixing conflicts - restoring newlines and indentation

Merge branch 'Add-Custom-Type-Annotation' of github.com:d3b-center/annoFuse into Add-Custom-Type-Annotation

# Conflicts:
#	R/fusion_standardization.R
#	man/fusion_standardization.Rd

federicomarini commented:
I had to restore some indentation, but otherwise to me it looks ready to merge

federicomarini commented:
merging done via CLI 🎉

sakshamphul commented:
Thank you @federicomarini for all the reviews, commits and merging this branch.

sakshamphul commented:
Hi @federicomarini How can we push these changes to CRAN webserver (release latest version)? Do we need to push to CRAN webserver?

federicomarini commented:
At the moment, this is not on CRAN or Bioc. So the installation has to go via Github, as it was before our tiny extension.
The plan to bring this to Bioc is alive and running, and it was mainly the motivation to split up the package in code (this) and data (annoFuseData).
Whether to proceed to Bioc would likely involve mainly @jharenza (decision-wise).

migbro commented:
@federicomarini @sakshamphul , now that the changes have been merged into master, I strongly suggest a release be made, following the guidelines here: https://semver.org/, so probably `v0.90.0` -> `v0.91.0`. It'd be a nice tidy way to allow users to use their preferred version easily as well as get a concise summary of what has changed since the last merge. Please let me know if you have any questions or comments. Thanks!

federicomarini commented:
Good point, also since we have the news.md file to keep track of this.
At some point we would (have to) jump to 0.99.0 if we enter the process of the Bioc submission, happy to oversee that if we want to go that way.
@sakshamphul, do you want to take over in a separate branch to update the NEWS and bump up the version number?

sakshamphul commented:
Sure! Just to confirm, latest version would be 0.91.0 ?

federicomarini commented:
Yeah, we are fairly free on this but that would be a reasonable choice. 
The x.y.Z number would be what goes up with tiny devel steps (e.g. in one release cycle of 6 months as on Bioc).
So, yes, go with 0.91.0
